mountpoint-s3/tests/fuse_tests/rmdir_test.rs

```diff
@@ -0,0 +1,66 @@
+use fuser::BackgroundSession;
+use mountpoint_s3::S3FilesystemConfig;
+use std::{
+    fs::{self, DirBuilder},
+    io::ErrorKind,
+};
+use tempfile::TempDir;
+use test_case::test_case;
+
+use crate::fuse_tests::{read_dir_to_entry_names, TestClientBox};
+
+fn rmdir_test<F>(creator_fn: F, prefix: &str)
+where
+    F: FnOnce(&str, S3FilesystemConfig) -> (TempDir, BackgroundSession, TestClientBox),
+{
+    let (mount_point, _session, _test_client) = creator_fn(prefix, Default::default());
+
+    // Create local directory
+    let main_dirname = "test_dir";
+    let main_path = mount_point.path().join(main_dirname);
+    let empty_dirname = "local_empty_dir";
+    let non_empty_dirname = "local_non_empty_dir";
+    let empty_dirpath = main_path.join(empty_dirname);
+    let non_empty_dirpath = main_path.join(non_empty_dirname);
+
+    DirBuilder::new().recursive(true).create(&empty_dirpath).unwrap();
+    DirBuilder::new().recursive(true).create(&non_empty_dirpath).unwrap();
+
+    // verify that the directories exist before testing `rmdir`
+    let read_dir_iter = fs::read_dir(&main_path).unwrap();
+    let dir_entry_names = read_dir_to_entry_names(read_dir_iter);
+    assert_eq!(dir_entry_names, vec![empty_dirname, non_empty_dirname]);
+
+    // Write an object into the directory
+    let filename = "nested_file";
+    let filepath = non_empty_dirpath.join(filename);
+    fs::write(filepath, "").unwrap();
+
+    // remove the directories
+    fs::remove_dir(&empty_dirpath).expect("should be able to remove empty directory");
+    //should not remove non empty directory
+    let err = fs::remove_dir(&non_empty_dirpath).expect_err("removing non-empty directory should fail");
+    assert_eq!(err.kind(), ErrorKind::PermissionDenied);
```

Comment:
I think we should check that `err.raw_os_error() == Some(libc::ENOTEMPTY)` instead of this

Reply:
Sure. Just wanted to know the reason. Is it because `ErrorKind` is not consistent across different OS?

Reply:
`ErrorKind` is consistent across OSes, which is kind of the problem—it needs to map error codes from many different platforms into a single enum. We want to test more directly that we're returning the `ENOTEMPTY` we expect, and not `EPERM` or some other error that also gets mapped to `PermissionDenied`.

There's a proposed Rust feature to make these errors more precise, and it includes [`DirectoryNotEmpty`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.DirectoryNotEmpty), but those haven't been stabilized yet so we can't use them.

Reply:
Oh! So, multiple os_error can be mapped to single ErrorKind. And we want to be correct about os error.